src/analytical_solutions/noh_solution/main.cpp

```diff
@@ -0,0 +1,219 @@
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+/*! @file
+ *
+ * @brief This program generate the analytical noh solution based in the time and the initial conditions
+ *
+ * @author Jose A. Escartin <ja.escartin@gmail.com>
+ *
+ */
+
+#include <string>
+#include <sstream>
+#include <iostream>
+#include <cmath>
+#include <vector>
+#include <filesystem>
+#include <algorithm>
+
+#include "arg_parser.hpp"
+#include "tests/particle_io.hpp"
+
+#include "noh/NohDataGenerator.hpp"
+
+#include "noh_io.hpp"
+#include "noh_solution.hpp"
+
+using namespace std;
+using namespace sphexa;
+
+using Real    = double;
+using KeyType = uint64_t;
+
+void printHelp(char* binName);
+
+int main(int argc, char** argv)
+{
+    const ArgParser parser(argc, argv);
+
+    if (parser.exists("-h") || parser.exists("--h") || parser.exists("-help") || parser.exists("--help"))
+    {
+        printHelp(argv[0]);
+        exit(EXIT_SUCCESS);
+    }
+
+    // Get command line parameters
+    const bool   only_sol  = parser.exists("--only_solution") ? true : false;
+    const double time      = parser.getDouble( "--time",   0.                  );
+    const size_t nParts    = parser.getInt(    "--nParts", 0                   );
+    const string inputFile = parser.getString( "--input",  "./dump_noh0.txt"   );
+    const string outDir    = parser.getString( "--outDir", "./"                );
+    const bool   complete  = parser.exists("--complete") ? true : false;
+
+    // Initialize vector size
+    vector<double> x  (nParts);
+    vector<double> y  (nParts);
+    vector<double> z  (nParts);
+    vector<double> vx (nParts);
+    vector<double> vy (nParts);
+    vector<double> vz (nParts);
+    vector<double> h  (nParts);
+    vector<double> rho(nParts);
+    vector<double> u  (nParts);
+    vector<double> p  (nParts);
+    vector<double> cs (nParts);
+    vector<double> Px (nParts);
+    vector<double> Py (nParts);
+    vector<double> Pz (nParts);
+    vector<ParticleIO> vSim(nParts);
```

Comment:
Why duplicate all these fields when you could just use `ParticlesData`? The data you're reading was created by a code that operated on `ParticlesData`, so if anything changes there, you'll have reflect the changes here (e.g. there's no guarantee that `double` will always be used for all fields).

Reply:
ParticleData integrated, only last sort by radius